File: gjs/mainloop.cpp
/* -*- mode: C++; c-basic-offset: 4; indent-tabs-mode: nil; -*- */
// SPDX-License-Identifier: MIT OR LGPL-2.0-or-later
// SPDX-FileCopyrightText: 2021 Evan Welsh <contact@evanwelsh.com>

#include <config.h>

#include <glib.h>

#include "gjs/auto.h"
#include "gjs/context-private.h"
#include "gjs/mainloop.h"

namespace Gjs {

bool MainLoop::spin(GjsContextPrivate* gjs) {
    if (m_exiting)
        return false;

    // Check if System.exit() has been called.
    if (gjs->should_exit(nullptr)) {
        // Return false to indicate the loop is exiting due to an exit call,
        // the queue is likely not empty
        debug("Not spinning loop because System.exit called");
        exit();
        return false;
    }

    Gjs::AutoPointer<GMainContext, GMainContext, g_main_context_unref>
        main_context{g_main_context_ref_thread_default()};

    debug("Spinning loop until released or hook cleared");
    do {
        bool blocking = can_block();

        // Only run the loop if there are pending jobs.
        if (g_main_context_pending(main_context))
            g_main_context_iteration(main_context, blocking);

        // If System.exit() has not been called
        if (gjs->should_exit(nullptr)) {
            debug("Stopped spinning loop because System.exit called");
            exit();
            return false;
        }
    } while (
        // and there is not a pending main loop hook
        !gjs->has_main_loop_hook() &&
        // and there are pending sources or the job queue is not empty
        // continue spinning the event loop.
        (can_block() || !gjs->empty()));

    return true;
}

};  // namespace Gjs

File: modules/esm/mainloop.js
// SPDX-License-Identifier: MIT OR LGPL-2.0-or-later
// SPDX-FileCopyrightText: 2022 Evan Welsh <contact@evanwelsh.com>

import GLib from 'gi://GLib';

const {idle_source} = imports.mainloop;

export const mainloop = GLib.MainLoop.new(null, false);

/**
 * Run the mainloop asynchronously
 *
 * @returns {Promise<void>}
 */
export function run() {
    return mainloop.runAsync();
}

/**
 * Quit the mainloop
 */
export function quit() {
    if (!mainloop.is_running())
        throw new Error('Main loop was stopped already');

    mainloop.quit();
}

/**
 * Adds an idle task to the mainloop
 *
 * @param {(...args) => boolean} handler a callback to call when no higher priority events remain
 * @param {number} [priority] the priority to queue the task with
 */
export function idle(handler, priority) {
    const source = idle_source(handler, priority);
    source.attach(null);

    return source;
}

export default {
    mainloop,
    idle,
    run,
    quit,
};
